domain/application/service/application.go

```diff
@@ -190,42 +206,84 @@ func makeAddUnitArgs(in AddUnitArg) application.AddUnitArg {
 // AddUnits adds the specified units to the application, returning an error
 // satisfying [applicationerrors.ApplicationNotFoundError] if the application doesn't exist.
 func (s *ApplicationService) AddUnits(ctx context.Context, name string, units ...AddUnitArg) error {
-	args := make([]application.AddUnitArg, len(units))
+	args := make([]application.UpsertUnitArg, len(units))
 	for i, u := range units {
-		args[i] = application.AddUnitArg{
+		args[i] = application.UpsertUnitArg{
 			UnitName: u.UnitName,
 		}
 	}
 	err := s.st.AddUnits(ctx, name, args...)
 	return errors.Annotatef(err, "adding units to application %q", name)
 }
 
-// UpsertCAASUnit creates or updates the specified application unit in a caas model,
+// RegisterCAASUnit creates or updates the specified application unit in a caas model,
 // returning an error satisfying [applicationerrors.ApplicationNotFoundError]
-// if the application doesn't exist.
-func (s *ApplicationService) UpsertCAASUnit(ctx context.Context, appName string, unit UpsertCAASUnitParams) error {
+// if the application doesn't exist. If the application life is Dead, an error
+// satisfying [applicationerrors.ApplicationIsDead] is returned.
+func (s *ApplicationService) RegisterCAASUnit(ctx context.Context, appName string, args RegisterCAASUnitParams) error {
+	if args.PasswordHash == nil {
+		return errors.NotValidf("password hash")
+	}
+	if args.ProviderId == nil {
+		return errors.NotValidf("provider id")
+	}
+	if !args.OrderedScale {
+		return errors.NewNotImplemented(nil, "registering CAAS units not supported without ordered unit IDs")
+	}
+	if args.UnitName == "" {
+		return errors.NotValidf("missing unit name")
+	}
+
 	p := AddUnitArg{
-		UnitName:     &unit.UnitName,
-		PasswordHash: unit.PasswordHash,
+		UnitName:     &args.UnitName,
+		PasswordHash: args.PasswordHash,
+		CloudContainer: &CloudContainerParams{
+			ProviderId: args.ProviderId,
+			Ports:      args.Ports,
+		},
 	}
-	if unit.ProviderId != nil || unit.Address != nil || unit.Ports != nil {
-		cldContainer := &CloudContainerParams{
-			ProviderId: unit.ProviderId,
-			Ports:      unit.Ports,
+	if args.Address != nil {
+		addr := network.NewSpaceAddress(*args.Address, network.WithScope(network.ScopeMachineLocal))
+		// k8s doesn't support spaces yet.
+		addr.SpaceID = network.AlphaSpaceId
+		p.CloudContainer.Address = &addr
+		origin := network.OriginProvider
+		p.CloudContainer.AddressOrigin = &origin
+	}
+	// We need to do a bunch of business logic in the one transaction so pass in a closure that is
+	// given the transaction to use.
+	unitArg := makeUpsertUnitArgs(p)
+	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		appID, err := s.st.GetApplicationID(ctx, appName)
```

Comment:
So here, RunAtomic is creating a TXN, and then we call GetApplicationID but it is actually just passing the AtomicContext as a plain context.Context, and looking at the code:
```
        var appID coreapplication.ID
        err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
                var err error
                appID, err = st.lookupApplication(ctx, tx, name, false)
                if err != nil {
                        return fmt.Errorf("looking up application %q: %w", name, err)
                }
                return nil
        })
```
It is fairly clear that it is creating an independent transaction.
At least, this is inefficient. I don't think the ApplicationID is changing so often that it matters, but this feels like a misdesign.